File: python/hash_file.py
#!/usr/bin/env python

import argparse
import csv
import re
import hashlib
import sys


debug = False

parser = argparse.ArgumentParser(description='Hash the contents of a file for TA upload.')

parser.add_argument('--type', required=True, metavar='TWITTERID', help='source data type.',
    choices=['MOBILEDEVICEID', 'IDFA', 'ADID', 'ANDROID', 'EMAIL', 'PHONE', 'TWITTERID', 'TWITTERSCREENNAME'])

parser.add_argument('--infile', required=True, type=argparse.FileType('rU'), metavar='/path/to/source.txt', help='input file to parse.')

parser.add_argument('--outfile', required=True, type=argparse.FileType('w'), metavar='/path/to/output.txt', help='file to write output.')

args = parser.parse_args()

flags = {'uppercase': False, 'dropleadingzeros': False, 'dropleadingat': False }

if args.type == 'MOBILEDEVICEID':
    # mobile device IDs can be a mixture of IDFA, ADID and ANDROID in a single file
    flags['regex'] = re.compile('^[a-z0-9][a-z0-9\-]+[a-z0-9]$')
elif args.type == 'IDFA':
    # flags['uppercase'] = True
    flags['regex'] = re.compile('^[a-z0-9][a-z0-9\-]+[a-z0-9]$')
elif args.type == 'ADID':
    flags['regex'] = re.compile('^[a-z0-9][a-z0-9\-]+[a-z0-9]$')
elif args.type == 'ANDROID':
    flags['regex'] = re.compile('^[a-z0-9]+$')
elif args.type == 'EMAIL':
    flags['regex'] = re.compile('^[a-z0-9][a-z0-9_\-\.\+]+\@[a-z0-9][a-z0-9\.]+[a-z]$')
elif args.type == 'PHONE' or args.type == 'TWITTERID':
    flags['dropleadingzeros'] = True
    flags['regex'] = re.compile('^\d+$')
elif args.type == 'TWITTERSCREENNAME':
    flags['dropleadingat'] = True
    flags['regex'] = re.compile('^[a-z0-9_]+$')
else:
    print ("ERROR: invalid type")
    sys.exit()

skipped = 0
written = 0

if args.infile.name.endswith(".csv"):
    csv_file = True
    reader = csv.reader(args.infile, dialect='excel')
else:
    csv_file = False
    reader = args.infile

for text in reader:
    if not csv_file:
        text = [text]

    for line in text:

        if not line: break

        line = line.rstrip()

        # Remove whitespace
        line = ''.join(line.split())

        # Set case
        if flags['uppercase']:
            line = line.upper()
        else:
            line = line.lower()

        # Drop leading '@'
        if flags['dropleadingat']:
            line = line.lstrip('@')

        # Drop leading zeros
        if flags['dropleadingzeros']:
            line = line.lstrip('0')

        if flags['regex'].match(line) is None:
            skipped += 1
            continue

        if debug:
            print ("\t" + line)

        hashed = hashlib.sha256(line).hexdigest()
        args.outfile.write(hashed + "\n")
        written += 1

args.infile.close()
args.outfile.close()

print ("Written:\t" + str(written))
print ("Skipped:\t" + str(skipped))

sys.exit()

File: perl/hash_file.pl
#!/usr/bin/perl

use strict;
use warnings;
use Digest::SHA qw(sha256_hex);
use File::Basename;

our $debug = 1;

# MAIN
{
    my ($type, $file) = @ARGV;

    unless ($file && -e $file) {
        die "File missing or does not exist\n";
    }

    unless ($type && $type =~ /^[A-Za-z]+$/) {
        die "Type is required\n";
    } else {
        $type = uc($type);
    }

    my ($file_name, $file_path, $file_suffix) = fileparse($file, qr/\.(txt|csv)/);

    unless ($file_suffix) {
        die "Invalid file type. Must be .txt or .csv, but got \"$file_suffix\"\n";
    }

    my $output_file;
    if ($file_name && $file_path && $file_suffix) {
        $output_file = $file_path . $file_name . "_hashed_$$" . $file_suffix;
    } else {
        die "Cannot format output file name\n";
    }

   my $flags = {};

    if ($type eq 'MOBILEDEVICEID') {
        # mobile device IDs can be a mixture of IDFA, ADID and ANDROID in a single file
        $flags->{'regex'} = qr/^[a-z0-9][a-z0-9\-]+[a-z0-9]$/;
    } elsif ($type eq 'IDFA') {
        $flags->{'regex'} = qr/^[a-z0-9][a-z0-9\-]+[a-z0-9]$/;
        # $flags->{'uppercase'} = 1;

    } elsif ($type eq 'ADID') {
        $flags->{'regex'} = qr/^[a-z0-9][a-z0-9\-]+[a-z0-9]$/;

    } elsif ($type eq 'ANDROID') {
        $flags->{'regex'} = qr/^[a-z0-9]+$/;

    } elsif ($type eq 'EMAIL') {
        $flags->{'regex'} = qr/^[a-z0-9][a-z0-9_\-\.\+]+\@[a-z0-9][a-z0-9\.]+[a-z]$/;

    } elsif ($type eq 'PHONE' || $type eq 'TWITTERID') {
        $flags->{'regex'} = qr/^\d+$/;
        $flags->{'dropleadingzeros'} = 1;

    } elsif ($type eq 'TWITTERSCREENNAME') {
        $flags->{'regex'} = qr/^[a-z0-9_]+$/;
        $flags->{'dropleadingat'} = 1;

    } else {
        die "Unknown type\n";
    }

    unless (open(INPUT, "<$file")) {
        die "Cannot open input file $file\n";
    }

    unless (open(OUTPUT, ">$output_file")) {
        die "Cannot open output file\n";
    }

    my $skipped = 0;
    my $written = 0;

    while (my $line = <INPUT>) {
        chomp($line);
        $line =~ s/\s//g;

        # Set case
        $line = $flags->{'uppercase'} ? uc($line) : lc($line);
        # Drop leading '@'
        $line =~ s/^\@// if ($flags->{'dropleadingat'});
        # Drop leading zeros
        $line =~ s/^0+// if ($flags->{'dropleadingzeros'});

        unless ($line =~ /$flags->{'regex'}/) {
            $skipped++;
            next;
        }

        if ($debug) {
            print "\t";
            print $line;
            print "\n";
        }
        print OUTPUT sha256_hex($line);
        print OUTPUT "\n";
        $written++;
    }

    close(INPUT);
    close(OUTPUT);

    print "Written:\t$written\n";
    print "Skipped:\t$skipped\n";
    exit;
}
